src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1616,6 +1616,81 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
 
 //------------------------------is_counted_loop--------------------------------
 bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  return do_is_counted_loop(x, loop, iv_bt) || is_counted_loop_with_speculative_long_limit(x, loop, iv_bt);
+}
+
+// When comparing an int incrementor with long limit, check if the limit is within int range. If so, treat it as an int
+// counted loop but deoptimize if the limit is out of int range.
+// This is common pattern with "for ( int i =...; i < long_limit; ...)" where int "i" is implicitly promoted to long,
+// "(long) i < long_limit", and therefore making it not an int counted loop without this optimization.
+bool PhaseIdealLoop::is_counted_loop_with_speculative_long_limit(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  if (iv_bt != T_INT) {
+    return false;
+  }
+
+  Node* back_control = loop_exit_control(x, loop);
+  if (back_control == nullptr) {
+    return false;
+  }
+
+  Node* ctrl = x->in(LoopNode::EntryControl);
+  if (!ctrl->is_IfTrue() || !ctrl->in(0)->is_ParsePredicate()) { // impossible to insert loop limit check?
+    return false;
+  }
+
+  BoolTest::mask bt = BoolTest::illegal;
+  float cl_prob = 0;
+  Node* incr = nullptr;
+  Node* limit = nullptr;
+  Node* cmp = loop_exit_test(back_control, loop, incr, limit, bt, cl_prob);
+
+  // Check exit test is a "(long) i < limit" pattern.
+  if (cmp == nullptr || cmp->Opcode() != Op_CmpL || incr == nullptr || incr->Opcode() != Op_ConvI2L) {
+    return false;
+  }
+
+  Node* new_incr = incr->in(1);
+
+  // Optimistically transform "(long) i < limit" to "i < (int) limit".
+  Node* new_limit = _igvn.register_new_node_with_optimizer(new ConvL2INode(limit), limit);
+  set_early_ctrl(new_limit, ctrl);
+
+  Node* new_cmp = _igvn.register_new_node_with_optimizer(
+      cmp->in(1) == incr ? new CmpINode(new_incr, new_limit) : new CmpINode(new_limit, new_incr),
+//      new CmpINode(new_incr, new_limit),
+      cmp);
+  set_early_ctrl(new_cmp, ctrl);
+
+  // back_control[0] -> iff[1] -> bool[1] -> cmp
+  Node* bol = back_control->in(0)->in(1);
+
+  _igvn.rehash_node_delayed(bol);
+  bol->replace_edge(cmp, new_cmp, &_igvn);
```

Comment:
You should use: `PhaseIdealLoop::insert_loop_limit_check_predicate()`. `do_is_counted_loop()` may need to add more loop limit checks and the way you've implemented it, I think you're replacing the place holder predicate that we use to create other ones.

Reply:
Actually you're not checking that it's a loop limit check so you could be replacing something else. You need to follow the pattern used elsewhere i.e.: https://github.com/openjdk/jdk/blob/2beb2b602bf20f1ec36e6244eca1a2eb50baccb4/src/hotspot/share/opto/loopnode.cpp#L2011

Reply:
FTR, we discussed this offline and I was looking at the wrong location.